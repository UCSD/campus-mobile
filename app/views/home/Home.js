import React from 'react'
import { Image, ScrollView, Alert } from 'react-native'
import { connect } from 'react-redux'
import { checkGooglePlayServices } from 'react-native-google-api-availability-bridge'

import WeatherCardContainer from '../weather/WeatherCardContainer'
import ShuttleCardContainer from '../shuttle/ShuttleCardContainer'
import EventCardContainer from '../events/EventCardContainer'
import QuicklinksCardContainer from '../quicklinks/QuicklinksCardContainer'
import NewsCardContainer from '../news/NewsCardContainer'
import DiningCardContainer from '../dining/DiningCardContainer'
import SpecialEventsCardContainer from '../specialEvents/SpecialEventsCardContainer'
import FinalsCard from '../schedule/FinalsCard'
import ScheduleCardContainer from '../schedule/ScheduleCardContainer'
import ParkingCardContainer from '../parking/ParkingCardContainer'
import { platformAndroid, gracefulFatalReset } from '../../util/general'
import css from '../../styles/css'
import logger from '../../util/logger'

const campusLogoImage = require('../../assets/images/UCSanDiegoLogo-nav.png')

export class Home extends React.Component {
	static navigationOptions = {
		title: 'Home',
		headerTitle: <Image source={campusLogoImage} style={css.navCampusLogoTitle} />
	};

	componentWillMount() {
		if (platformAndroid()) {
			this.props.user.updatedGoogle = true
			this.updateGooglePlay()
		}
	}

	componentDidMount() {
		logger.ga('View Loaded: Home')
		this._cards = []

		if (this._scrollview) {
			this._scrollview.scrollTo({ y: this.props.lastScroll, animated: false })
		}
	}

	shouldComponentUpdate(nextProps, nextState) {
		if (this.props.cards !== nextProps.cards ||
			this.props.cardOrder !== nextProps.cardOrder ||
			this.props.user.isLoggedIn !== nextProps.user.isLoggedIn) {
			return true
		} else {
			return false
		}
	}

	componentDidUpdate(prevProps) {
		if (!prevProps.user.invalidSavedCredentials &&
			this.props.user.invalidSavedCredentials) {
			Alert.alert(
				'Logged Out.',
				'You have been logged out because your credentials could not be verified. Please try to log in again.',
				[
					{
						text: 'Cancel',
						style: 'cancel'
					},
					{
						text: 'Log in',
						onPress: () => {
							this.props.navigation.navigate('LoginScreen')
						}
					}
				]
			)
		}
	}

	handleScroll = (event) => {
		if (this.props.updateScroll) {
			this.props.updateScroll(event.nativeEvent.contentOffset.y)
		}
	}

	_getCards = () => {
		const activeCards = []

		if (Array.isArray(this.props.cardOrder)) {
			this.props.cardOrder.forEach((card) => {
				if (this.props.cards[card].active) {
					// Skip cards if they require authentication
					// and user is not authenticated
					if (this.props.cards[card].authenticated) {
						if (!this.props.user.isLoggedIn) return
						else if (this.props.cards[card].classifications
							&& this.props.cards[card].classifications.student
							&& !this.props.user.profile.classifications.student) return
					}

					switch (card) {
						case 'specialEvents':
							activeCards.push(<SpecialEventsCardContainer key="specialEvents" />)
							break
						case 'finals':
							activeCards.push(<FinalsCard key="finals" />)
							break
						case 'schedule':
							activeCards.push(<ScheduleCardContainer key="schedule" />)
							break
						case 'weather':
							activeCards.push(<WeatherCardContainer key="weather" />)
							break
						case 'shuttle':
							activeCards.push(<ShuttleCardContainer key="shuttle" />)
							break
						case 'dining':
							activeCards.push(<DiningCardContainer key="dining" />)
							break
						case 'events':
							activeCards.push(<EventCardContainer key="events" />)
							break
						case 'quicklinks':
							activeCards.push(<QuicklinksCardContainer key="quicklinks" />)
							break
						case 'news':
							activeCards.push(<NewsCardContainer key="news" />)
							break
						case 'parking':
							activeCards.push(<ParkingCardContainer key="parking" />)
							break
						default:
							return gracefulFatalReset(new Error('Invalid card in state: ', card))
					}
				}
			})
		}
		return activeCards
	}

	updateGooglePlay = () => {
		checkGooglePlayServices((result) => {
			if (result === 'update') {
				this.props.user.updatedGoogle = false
			}
		})
	}

	render() {
		return (
			<ScrollView
				ref={(c) => { this._scrollview = c }}
				onScroll={this.handleScroll}
				scrollEventThrottle={0}
			>
				{/* LOAD CARDS */}
				{ this._getCards() }
			</ScrollView>
		)
	}
}

function mapStateToProps(state, props) {
	return {
		cards: state.cards.cards,
		cardOrder: state.cards.cardOrder,
		lastScroll: state.home.lastScroll,
		user: state.user
	}
}

function mapDispatchtoProps(dispatch) {
	return {
		updateScroll: (scrollY) => {
			dispatch({ type: 'UPDATE_HOME_SCROLL', scrollY })
		}
	}
}
export default connect(mapStateToProps, mapDispatchtoProps)(Home)
