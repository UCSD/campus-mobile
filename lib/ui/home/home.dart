import 'dart:async';

import 'package:campus_mobile_experimental/app_constants.dart';
import 'package:campus_mobile_experimental/app_styles.dart';
import 'package:campus_mobile_experimental/core/models/cards.dart';
import 'package:campus_mobile_experimental/core/models/notices.dart';
import 'package:campus_mobile_experimental/core/providers/bottom_nav.dart';
import 'package:campus_mobile_experimental/core/providers/cards.dart';
import 'package:campus_mobile_experimental/core/providers/map.dart';
import 'package:campus_mobile_experimental/core/providers/notices.dart';
import 'package:campus_mobile_experimental/core/providers/wayfinding.dart';
import 'package:campus_mobile_experimental/main.dart';
import 'package:campus_mobile_experimental/ui/availability/availability_card.dart';
import 'package:campus_mobile_experimental/ui/classes/classes_card.dart';
import 'package:campus_mobile_experimental/ui/common/webview_container.dart';
import 'package:campus_mobile_experimental/ui/dining/dining_card.dart';
import 'package:campus_mobile_experimental/ui/employee_id/employee_id_card.dart';
import 'package:campus_mobile_experimental/ui/events/events_card.dart';
import 'package:campus_mobile_experimental/ui/finals/finals_card.dart';
import 'package:campus_mobile_experimental/ui/my_chart/my_chart_card.dart';
import 'package:campus_mobile_experimental/ui/myucsdchart/myucsdchart.dart';
import 'package:campus_mobile_experimental/ui/navigator/top.dart';
import 'package:campus_mobile_experimental/ui/news/news_card.dart';
import 'package:campus_mobile_experimental/ui/notices/notices_card.dart';
import 'package:campus_mobile_experimental/ui/parking/parking_card.dart';
import 'package:campus_mobile_experimental/ui/scanner/native_scanner_card.dart';
import 'package:campus_mobile_experimental/ui/shuttle/shuttle_card.dart';
import 'package:campus_mobile_experimental/ui/student_id/student_id_card.dart';
import 'package:campus_mobile_experimental/ui/survey/survey_card.dart';
import 'package:campus_mobile_experimental/ui/ventilation/ventilation_card.dart';
import 'package:campus_mobile_experimental/ui/weather/weather_card.dart';
import 'package:campus_mobile_experimental/ui/wifi/wifi_card.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:uni_links2/uni_links.dart';

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  Future<Null> initUniLinks(BuildContext context) async {
    // deep links are received by this method
    // the specific host needs to be added in AndroidManifest.xml and Info.plist
    // currently, this method handles executing custom map query
    late StreamSubscription _sub;

    // Used to handle links on cold app start
    String? initialLink = await getInitialLink();
    if (!executedInitialDeeplinkQuery &&
        initialLink != null &&
        initialLink.contains("deeplinking.searchmap")) {
      var uri = Uri.dataFromString(initialLink);
      var query = uri.queryParameters['query']!;
      // redirect query to maps tab and search with query
      executeQuery(context, query);
    }

    // used to handle links while app is in foreground/background
    _sub = linkStream.listen((String? link) async {
      // handling for map query
      if (link!.contains("deeplinking.searchmap")) {
        var uri = Uri.dataFromString(link);
        var query = uri.queryParameters['query']!;
        // redirect query to maps tab and search with query
        executeQuery(context, query);
        // received deeplink, cancel stream to prevent memory leaks
        _sub.cancel();
      }
    });
  }

  void executeQuery(BuildContext context, String query) {
    Provider.of<MapsDataProvider>(context, listen: false)
        .searchBarController
        .text = query;
    Provider.of<MapsDataProvider>(context, listen: false).fetchLocations();
    Provider.of<BottomNavigationBarProvider>(context, listen: false)
        .currentIndex = NavigatorConstants.MapTab;
    Provider.of<CustomAppBar>(context, listen: false).changeTitle("Maps");
    executedInitialDeeplinkQuery = true;
  }

  @override
  Widget build(BuildContext context) {
    initUniLinks(context);
    checkToResumeBluetooth(context);
    return SingleChildScrollView(
      child: Padding(
        padding: EdgeInsets.only(
            top: cardMargin + 2.0,
            right: cardMargin,
            bottom: 0.0,
            left: cardMargin),
        child: Column(
          children: createList(context),
        ),
      ),
    );
  }

  List<Widget> createList(BuildContext context) {
    List<Widget> orderedCards =
        getOrderedCardsList(Provider.of<CardsDataProvider>(context).cardOrder!);
    List<Widget> noticesCards = getNoticesCardsList(
        Provider.of<NoticesDataProvider>(context).noticesModel!);

    return noticesCards + orderedCards;
  }

  List<Widget> getNoticesCardsList(List<NoticesModel> notices) {
    List<Widget> noticesCards = [];
    for (NoticesModel notice in notices) {
      noticesCards.add(NoticesCard(notice: notice));
    }
    return noticesCards;
  }

  List<Widget> getOrderedCardsList(List<String> order) {
    List<Widget> orderedCards = [];
    Map<String, CardsModel?>? webCards =
        Provider.of<CardsDataProvider>(context, listen: false).webCards;

    for (String card in order) {
      if (!webCards!.containsKey(card)) {
        switch (card) {
          case 'NativeScanner':
            orderedCards.insert(0, NativeScannerCard());
            break;
          case 'MyStudentChart':
            orderedCards.add(MyStudentChartCard());
            break;
          case 'student_survey':
            orderedCards.add(SurveyCard());
            break;
          case 'dining':
            orderedCards.add(DiningCard());
            break;
          case 'news':
            orderedCards.add(NewsCard());
            break;
          case 'events':
            orderedCards.add(EventsCard());
            break;
          case 'weather':
            orderedCards.add(WeatherCard());
            break;
          case 'availability':
            orderedCards.add(AvailabilityCard());
            break;
          case 'schedule':
            orderedCards.add(ClassScheduleCard());
            break;
          case 'finals':
            orderedCards.add(FinalsCard());
            break;
          case 'MyUCSDChart':
            orderedCards.add(MyUCSDChartCard());
            break;
          case 'student_id':
            orderedCards.add(StudentIdCard());
            break;
          case 'employee_id':
            orderedCards.add(EmployeeIdCard());
            break;
          case 'parking':
            orderedCards.add(ParkingCard());
            break;
          case 'speed_test':
            orderedCards.add(WiFiCard());
            break;
          case 'shuttle':
            orderedCards.add(ShuttleCard());
            break;
          case 'ventilation':
            orderedCards.add(VentilationCard());
            break;
        }
      } else {
        // dynamically insert webCards into the list
        orderedCards.add(WebViewContainer(
          titleText: webCards[card]!.titleText,
          initialUrl: webCards[card]!.initialURL,
          cardId: card,
          requireAuth: webCards[card]!.requireAuth,
        ));
      }
    }
    return orderedCards;
  }

  void checkToResumeBluetooth(BuildContext context) async {
    SharedPreferences prefs = await SharedPreferences.getInstance();
    bool? preference = prefs.getBool("advancedWayfindingEnabled");

    if (prefs.containsKey("advancedWayfindingEnabled") && preference == true) {
      WayfindingProvider bluetoothSingleton =
          Provider.of<WayfindingProvider>(context, listen: false);
      bluetoothSingleton.advancedWayfindingEnabled = true;
      if (bluetoothSingleton.ongoingScanner == null) {
        bluetoothSingleton.init();
      }
    }
  }
}
