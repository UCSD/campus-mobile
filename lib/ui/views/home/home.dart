import 'package:campus_mobile_experimental/core/data_providers/cards_data_provider.dart';
import 'package:campus_mobile_experimental/core/data_providers/notices_data_provider.dart';
import 'package:campus_mobile_experimental/core/models/notices_model.dart';
import 'package:campus_mobile_experimental/ui/cards/availability/availability_card.dart';
import 'package:campus_mobile_experimental/ui/cards/class_schedule/class_schedule_card.dart';
import 'package:campus_mobile_experimental/ui/cards/dining/dining_card.dart';
import 'package:campus_mobile_experimental/ui/cards/events/events_card.dart';
import 'package:campus_mobile_experimental/ui/cards/finals/finals_card.dart';
import 'package:campus_mobile_experimental/ui/cards/my_chart/my_chart_card.dart';
import 'package:campus_mobile_experimental/ui/cards/news/news_card.dart';
import 'package:campus_mobile_experimental/ui/cards/notices/notices_card.dart';
import 'package:campus_mobile_experimental/ui/cards/scanner/scanner_card.dart';
import 'package:campus_mobile_experimental/ui/cards/staff_id/staff_id_card.dart';
import 'package:campus_mobile_experimental/ui/cards/staff_info/staff_info_card.dart';
import 'package:campus_mobile_experimental/ui/cards/student_id/student_id_card.dart';
import 'package:campus_mobile_experimental/ui/cards/student_info/student_info_card.dart';
import 'package:campus_mobile_experimental/ui/cards/weather/weather_card.dart';
import 'package:campus_mobile_experimental/ui/cards/campus_info/campus_info_card.dart';
import 'package:campus_mobile_experimental/ui/theme/app_layout.dart';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: EdgeInsets.symmetric(horizontal: cardMargin, vertical: 0.0),
      child: ListView(
        padding: EdgeInsets.only(
            top: cardMargin + 2.0, right: 0.0, bottom: 0.0, left: 0.0),
        children: createList(context),
      ),
    );
  }

  List<Widget> createList(BuildContext context) {
    List<Widget> orderedCards =
        getOrderedCardsList(Provider.of<CardsDataProvider>(context).cardOrder);
    List<Widget> noticesCards = getNoticesCardsList(
        Provider.of<NoticesDataProvider>(context).noticesModel);

    return noticesCards + orderedCards;
  }

  List<Widget> getNoticesCardsList(List<NoticesModel> notices) {
    List<Widget> noticesCards = List<Widget>();
    for (NoticesModel notice in notices) {
      noticesCards.add(NoticesCard(notice: notice));
    }
    return noticesCards;
  }

  List<Widget> getOrderedCardsList(List<String> order) {
    List<Widget> orderedCards = List<Widget>();

    for (String card in order) {
      switch (card) {
        case 'QRScanner':
          orderedCards.insert(0, ScannerCard());
          break;
        case 'MyUCSDChart':
          orderedCards.add(MyChartCard());
          break;
        case 'staff_info':
          orderedCards.add(StaffInfoCard());
          break;
        case 'campus_info':
          orderedCards.add(CampusInfoCard());
          break;
        case 'student_info':
          orderedCards.add(StudentInfoCard());
          break;
        case 'student_id':
          orderedCards.add(StudentIdCard());
          break;
        case 'finals':
          orderedCards.add(FinalsCard());
          break;
        case 'schedule':
          orderedCards.add(ClassScheduleCard());
          break;
        case 'dining':
          orderedCards.add(DiningCard());
          break;
        case 'events':
          orderedCards.add(EventsCard());
          break;
        case 'news':
          orderedCards.add(NewsCard());
          break;
        case 'student_id':
          orderedCards.add(StudentIdCard());
          break;
        case 'staff_id':
          orderedCards.add(StaffIdCard());
          break;
//        case 'special_events':
//          orderedCards.add(BannerCard());
//          break;
        case 'weather':
          orderedCards.add(WeatherCard());
          break;
        case 'availability':
          orderedCards.add(AvailabilityCard());
          break;
      }
    }
    return orderedCards;
  }
}
