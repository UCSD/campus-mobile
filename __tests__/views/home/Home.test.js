import 'react-native'
import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

// Component to be tested
import { Home } from '../../../app/views/home/Home'

Enzyme.configure({ adapter: new Adapter() })

// Mock props passed down from state
const initialState = {
	cards: {
		specialEvents: {
			id: 'specialEvents',
			active: false,
			autoActivated: false,
			name: 'Special Events',
			component: 'SpecialEventsCard'
		},
		finals: {
			id: 'finals',
			active: true,
			name: 'Finals',
			component: 'FinalsCard',
			authenticated: true,
			autoActivated: false,
			classifications: { student: true }
		},
		schedule: {
			id: 'schedule',
			active: true,
			name: 'Classes',
			component: 'ScheduleCard',
			authenticated: true,
			classifications: { student: true }
		},
		shuttle: {
			id: 'shuttle',
			active: true,
			name: 'Shuttle',
			component: 'ShuttleCard'
		},
		dining: {
			id: 'dining',
			active: true,
			name: 'Dining',
			component: 'DiningCard'
		},
		events: {
			id: 'events',
			active: true,
			name: 'Events',
			component: 'EventsCard'
		},
		news: {
			id: 'news',
			active: true,
			name: 'News',
			component: 'NewsCard',
		},
		quicklinks: {
			id: 'quicklinks',
			active: true,
			name: 'Links',
			component: 'QuicklinksCard'
		},
		weather: {
			id: 'weather',
			active: true,
			name: 'Weather',
			component: 'WeatherCard'
		},
		parking: {
			id: 'parking',
			active: true,
			name: 'Parking',
			component: 'ParkingCard'
		}
	},

	// Only cards that show up by default
	// on first launch should appear here.
	cardOrder: [
		'shuttle',
		'parking',
		'dining',
		'events',
		'news',
		'quicklinks',
		'weather'
	],
}

// Set up component to be rendered
function setup(props) {
	return shallow(<Home {...props} />)
}

test('renders without crashing', () => {
	const tree = setup(initialState)
	expect(tree).toMatchSnapshot()
})
