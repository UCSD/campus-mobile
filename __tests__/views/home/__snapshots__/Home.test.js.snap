// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<ScrollView
  onScroll={[Function]}
  scrollEventThrottle={0}
>
  <Connect(withNavigation(ShuttleCardContainer))
    key="shuttle"
  />
  <Connect(withNavigation(ParkingCardContainer))
    key="parking"
  />
  <Connect(DiningCardContainer)
    key="dining"
  />
  <Connect(EventCardContainer)
    key="events"
  />
  <Connect(NewsCardContainer)
    key="news"
  />
  <Connect(QuicklinksCardContainer)
    key="quicklinks"
  />
  <Connect(WeatherCardContainer)
    key="weather"
  />
</ScrollView>
`;
